go/cmd/server/server.go

```diff
@@ -49,21 +52,87 @@ func (s *servingServiceServer) GetOnlineFeatures(ctx context.Context, request *s
 	}
+	// Entities are currently part of the features as a value and the order that we add it to the resp MetaData
+	// Need to figure out a way to map the correct entities to the correct ordering
+	entityValuesMap := make(map[string][]*prototypes.Value, 0)
+	featureNames := make([]string, len(featureVectors))
+	for idx, vector := range featureVectors {
 
-	for _, vector := range featureVectors {
 		resp.Metadata.FeatureNames.Val = append(resp.Metadata.FeatureNames.Val, vector.Name)
-
+		featureNames[idx] = vector.Name
 		values, err := types.ArrowValuesToProtoValues(vector.Values)
 		if err != nil {
 			return nil, err
 		}
-
+		if _, ok := request.Entities[vector.Name]; ok {
+			entityValuesMap[vector.Name] = values
+		}
 		resp.Results = append(resp.Results, &serving.GetOnlineFeaturesResponse_FeatureVector{
 			Values:          values,
 			Statuses:        vector.Statuses,
 			EventTimestamps: vector.Timestamps,
 		})
 	}
 
+	go generateLogs(s, entityValuesMap, featureNames, resp.Results, request.RequestContext)
 	return resp, nil
 }
+
+func generateLogs(s *servingServiceServer, entityMap map[string][]*prototypes.Value, featureNames []string, features []*serving.GetOnlineFeaturesResponse_FeatureVector, requestData map[string]*prototypes.RepeatedValue) error {
+	// Add a log with the request context
+	if len(requestData) > 0 {
+		requestContextLog := logging.Log{
+			RequestContext: requestData,
+		}
+		s.loggingService.EmitLog(&requestContextLog)
+	}
+
+	if len(features) <= 0 {
+		return nil
+	}
+
+	entityArr, err := s.fs.ListEntities(true)
+	if err != nil {
+		return err
+	}
+
+	joinKeys := make([]string, 0)
+	for _, entity := range entityArr {
+		joinKeys = append(joinKeys, entity.JoinKey)
+	}
+
+	numFeatures := len(featureNames)
+	// Should be equivalent to how many entities there are(each feature row has (entity) number of features)
+	// acc_rate []
+	numRows := len(features[0].Values)
+	featureValueLogRows := make([][]*prototypes.Value, numRows)
+	featureStatusLogRows := make([][]serving.FieldStatus, numRows)
+	eventTimestampLogRows := make([][]*timestamppb.Timestamp, numRows)
+
+	for row_idx := 0; row_idx < numRows; row_idx++ {
+		featureValueLogRows[row_idx] = make([]*prototypes.Value, numFeatures)
+		featureStatusLogRows[row_idx] = make([]serving.FieldStatus, numFeatures)
+		eventTimestampLogRows[row_idx] = make([]*timestamppb.Timestamp, numFeatures)
+		for idx := 1; idx < len(features); idx++ {
+			featureValueLogRows[row_idx][idx-1] = features[idx].Values[row_idx]
+			featureStatusLogRows[row_idx][idx-1] = features[idx].Statuses[row_idx]
+			eventTimestampLogRows[row_idx][idx-1] = features[idx].EventTimestamps[row_idx]
+		}
+		entityRow := make([]*prototypes.Value, 0)
+		// ensure that the entity values are in the order that the schema defines which is the order that ListEntities returns the entities
+		for _, joinKey := range joinKeys {
+			entityRow = append(entityRow, entityMap[joinKey][row_idx])
+		}
+		newLog := logging.Log{
```

Comment:
if log is being emitted per each row, we need to have request id (that is should probably be generated somewhere here) in each log to understand which rows came from which request

Reply:
Done.